cmd/akamai-purger/main.go

```diff
@@ -193,36 +194,54 @@ func (ap *akamaiPurger) len() int {
 
 func (ap *akamaiPurger) purge() error {
 	ap.Lock()
-	queueEntries := ap.toPurge[:]
-	ap.toPurge = [][]string{}
-	ap.Unlock()
-	if len(queueEntries) == 0 {
+	queueSize := len(ap.toPurge) - 1
+	batchSize := ap.queueEntriesPerBatch
+
+	// If the queue is empty, return immediately.
+	if queueSize <= 0 {
+		ap.Unlock()
 		return nil
 	}
 
-	stoppedAt, err := ap.client.Purge(queueEntries)
+	// If the queue contains less than full batch, set the batch size to the
+	// queue size.
+	if queueSize < batchSize {
+		batchSize = queueSize
+	}
+	batchBegin := queueSize - batchSize
+
+	purgeBatch := ap.toPurge[batchBegin:]
+	ap.toPurge = ap.toPurge[:batchBegin]
+	ap.Unlock()
+
+	err := ap.client.Purge(purgeBatch)
 	if err != nil {
 		ap.Lock()
-
-		// Add the remaining queue entries back, but at the end of the queue. If somehow
-		// there's a URL which repeatedly results in error, it won't block the
-		// entire queue, only a single batch.
-		ap.toPurge = append(ap.toPurge, queueEntries[stoppedAt:]...)
+		if queueSize > ap.maxQueueSize {
+			// Drop the oldest entry from the queue to make room.
+			ap.toPurge = ap.toPurge[1:]
+		}
+		// Add the entries from the failed batch to the back of the queue to
+		// avoid blocking the entire queue.
```

Comment:
A couple of notes here: this is no longer a queue, it's a stack (we should search the source files for "queue" and fix it).

Also, what should would do when appending failed items for later retry? Should they act like very "new" items and get pushed onto the top of the stack (high priority for retry)? Or should they act like very "old" items and get snuck[1] in on the bottom of the stack (low priority for retry)?

I actually think we should drop these on the floor. We've already concluded that dropping purger entries in situations of unusual load is okay, and we already have internal retries for a single purge batch.

[1]: I say "snuck" because conceptually a stack doesn't really support "put something on the bottom." Nor, for that matter, do they support "pop something off the bottom." So we're being a bit loose about treating this as a stack.

Reply:
I'm hesitant to drop these entries on the floor when we have plenty of memory to continue to store and re-attempt during periods of intermittent connectivity. As to the order, I don't believe that in most cases sending failed entries to the bottom of the stack will result in their being more than 2 - 3 batches away from another re-attempt. The current implementation sees its queue drained on nearly every tick with just a few exceptions to that noted each day.